app/src/main/java/com/duckduckgo/app/autocomplete/api/AutoComplete.kt

```diff
@@ -75,12 +78,69 @@ class AutoCompleteApi @Inject constructor(
             .toObservable()
 
     private fun getAutoCompleteBookmarkResults(query: String) =
-        bookmarksDao.bookmarksByQuery("%$query%")
+        bookmarksDao.bookmarksObservable()
+            .map { rankBookmarks(query, it) }
             .flattenAsObservable { it }
             .map {
-                AutoCompleteBookmarkSuggestion(phrase = it.url, title = it.title ?: "", url = it.url)
+                AutoCompleteBookmarkSuggestion(phrase = it.url.toUri().baseHost.orEmpty(), title = it.title.orEmpty(), url = it.url)
             }
+            .take(2)
             .toList()
             .onErrorReturn { emptyList() }
             .toObservable()
+
+    private fun rankBookmarks(query: String, bookmarks: List<BookmarkEntity>): List<BookmarkEntity> {
+        return bookmarks.asSequence()
+            .map { RankedBookmark(bookmarkEntity = it) }
+            .map { scoreTitle(it, query) }
+            .map { scoreTokens(it, query) }
+            .filter { it.score >= 0 }
+            .sortedBy { it.score }
+            .map { it.bookmarkEntity }
+            .toList()
+    }
+
+    private fun scoreTitle(rankedBookmark: RankedBookmark, query: String): RankedBookmark {
+        if (rankedBookmark.bookmarkEntity.title?.startsWith(query, ignoreCase = true) == true) {
+            rankedBookmark.score += 200
+        } else if (rankedBookmark.bookmarkEntity.title?.contains(" $query", ignoreCase = true) == true) {
+            rankedBookmark.score += 100
+        }
+
+        return rankedBookmark
+    }
+
+    private fun scoreTokens(rankedBookmark: RankedBookmark, query: String): RankedBookmark {
+        val bookmark = rankedBookmark.bookmarkEntity
+        val domain = bookmark.url.toUri().baseHost
+        val tokens = query.split(" ")
+        if (tokens.size > 1) {
+            tokens.forEach { token ->
+                if (bookmark.title?.startsWith(token, ignoreCase = true) == false &&
+                    bookmark.title?.contains(" $token", ignoreCase = true) == false &&
+                    domain?.startsWith(token, ignoreCase = true) == false
+                ) {
+                    return rankedBookmark
+                }
+            }
+
+            rankedBookmark.score += 10
+
+            if (domain?.startsWith(tokens.first(), ignoreCase = true) == true) {
+                rankedBookmark.score += 300
+            } else if (bookmark.title?.startsWith(tokens.first(), ignoreCase = true) == true) {
+                rankedBookmark.score += 50
+            }
+
+        } else if (domain?.startsWith(tokens.first(), ignoreCase = true) == true) {
+            rankedBookmark.score += 300
+        }
+
+        return rankedBookmark
+    }
+
+    private data class RankedBookmark(
+        val bookmarkEntity: BookmarkEntity,
+        var score: Int = -1
```

Comment:
nit: I know this is -1 because that's the iOS implementation and they have `0` for favourites. Maybe moving it to a const like `BOOKMARK_SCORE`? Your call, I'm not too fuzz about it.